src/Week1/HW01.hs

```diff
@@ -14,17 +14,38 @@ dropLastDigit n = n `div` 10
 
 -- Exercise 2 -----------------------------------------
 
+{- toRevDigits should convert positive Integers to a list of digits. (For
+0 or negative inputs, toRevDigits should return the empty list.) -}
+
+-- non tail recursive
 toRevDigits :: Integer -> [Integer]
-toRevDigits n = helper n []
+toRevDigits n 
+        | n <= 0 = [] 
+        | otherwise = lastDigit n : toRevDigits (dropLastDigit n)
+
+-- tail recursive version, assuming reverse is tail recursive
+toRevDigits' :: Integer -> [Integer]
+toRevDigits' n = helper n []
+    where
+        helper :: Integer -> [Integer] -> [Integer]
+        helper x acc
+            | x <= 0 = reverse acc
+            | otherwise = helper (dropLastDigit x) (lastDigit x : acc)
+
+-- tail recursive non reversed digits
+toDigits :: Integer -> [Integer]
+toDigits n = helper n []
     where
-        helper x acc 
-            | x <= 0 = reverse acc 
+        helper :: Integer -> [Integer] -> [Integer]
+        helper x acc
+            | x <= 0 = acc
             | otherwise = helper (dropLastDigit x) (lastDigit x : acc)
     
 
 -- Exercise 3 -----------------------------------------
 
 -- Double every second number in a list starting on the left.
+-- If only there was a mapi in haskell
```

Comment:
There is imap in various lens libraries, but in basic haskell you can achieve the same using [zipWith](https://hackage.haskell.org/package/base-4.15.0.0/docs/Prelude.html#v:zipWith)

```haskell
mapi :: (Int -> a -> b) -> [a] -> [b]
mapi f as = zipWith f [1..] as
```

Reply:
In the code below it's usually safer to use pattern matching than head and tail, which should be ideally avoided because they are partial and can fail at runtime.
Example:

bad, because it crashes at runtime
```haskell
f xs = doSomethingWith (head xs)   (tail xs) 
```

good, because compiler has your back (can check that you're handling all possible patterns via pattern matching)
```haskell
f [] = ... -- handle empty list
f (x:xs) = doSomethingWith x xs
```

Even if you're checking for list emptiness it's still a just runtime check.
Letting compiler check that you're handling all the cases is usually preferable.